lightning/src/offers/invoice_request.rs

```diff
@@ -0,0 +1,450 @@
+
+impl InvoiceRequest {
+	/// An unpredictable series of bytes, typically containing information about the derivation of
+	/// [`payer_id`].
+	///
+	/// [`payer_id`]: Self::payer_id
+	pub fn metadata(&self) -> Option<&Vec<u8>> {
+		self.contents.payer.0.as_ref()
+	}
+
+	/// A chain from [`Offer::chains`] that the offer is valid for.
+	///
+	/// [`Offer::chains`]: crate::offers::offer::Offer::chains
+	pub fn chain(&self) -> ChainHash {
+		self.contents.chain()
+	}
+
+	/// The amount to pay in msats (i.e., the minimum lightning-payable unit for [`chain`]), which
+	/// must be greater than or equal to [`Offer::amount`], converted if necessary.
+	///
+	/// [`chain`]: Self::chain
+	/// [`Offer::amount`]: crate::offers::offer::Offer::amount
+	pub fn amount_msats(&self) -> Option<u64> {
+		self.contents.amount_msats
+	}
+
+	/// Features for paying the invoice.
+	pub fn features(&self) -> Option<&OfferFeatures> {
+		self.contents.features.as_ref()
+	}
+
+	/// The quantity for the offer's item within the range [`Offer::quantity_min`] to
+	/// [`Offer::quantity_max`], inclusive.
+	///
+	/// [`Offer::quantity_min`]: crate::offers::offer::Offer::quantity_min
+	/// [`Offer::quantity_max`]: crate::offers::offer::Offer::quantity_max
+	pub fn quantity(&self) -> Option<u64> {
+		self.contents.quantity
+	}
+
+	/// A transient pubkey used to sign the invoice request.
+	pub fn payer_id(&self) -> PublicKey {
+		self.contents.payer_id
+	}
+
+	/// Payer provided note to include in the invoice.
+	pub fn payer_note(&self) -> Option<&String> {
+		self.contents.payer_note.as_ref()
+	}
+
+	/// Signature of the invoice request using [`payer_id`].
+	///
+	/// [`payer_id`]: Self::payer_id
+	pub fn signature(&self) -> Option<Signature> {
+		self.signature
+	}
+}
+
+impl InvoiceRequestContents {
+	fn chain(&self) -> ChainHash {
+		self.chain.unwrap_or_else(|| self.offer.implied_chain())
+	}
+
+	pub(super) fn as_tlv_stream(&self) -> PartialInvoiceRequestTlvStreamRef {
+		let payer = PayerTlvStreamRef {
+			metadata: self.payer.0.as_ref(),
+		};
+
+		let offer = self.offer.as_tlv_stream();
+
+		let invoice_request = InvoiceRequestTlvStreamRef {
+			chain: self.chain.as_ref(),
+			amount: self.amount_msats,
+			features: self.features.as_ref(),
+			quantity: self.quantity,
+			payer_id: Some(&self.payer_id),
+			payer_note: self.payer_note.as_ref(),
+		};
+
+		(payer, offer, invoice_request)
+	}
+}
+
+impl Writeable for InvoiceRequest {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		WithoutLength(&self.bytes).write(writer)
+	}
+}
+
+impl Writeable for InvoiceRequestContents {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		self.as_tlv_stream().write(writer)
+	}
+}
+
+impl TryFrom<Vec<u8>> for InvoiceRequest {
+	type Error = ParseError;
+
+	fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+		let tlv_stream: FullInvoiceRequestTlvStream = Readable::read(&mut &bytes[..])?;
+		InvoiceRequest::try_from((bytes, tlv_stream))
```

Comment:
@TheBlueMatt @valentinewallace 

Unfortunately, it turns out reading tuples of TLV stream structs doesn't work as expected as currently implemented. Reading the first TLV stream will attempt to exhaust the entire byte sequence and will fail when seeing the first even-typed TLV record from the second TLV stream.

Some possible solutions would be:
1. Duplicated the TLV records in each `tlv_stream` usage instead of using tuples of TLV stream structs (e.g., duplicate the `OfferTlvStream` records in `InvoiceRequestTlvStream`).
2. Parse the byte sequence into TLV record ranges `0`, `1..80`, `80-160`, `160-240`, `240..=1000`, `1001..` and read them into each TLV stream type separately
3. Modify the decode macros to understand type ranges

(1) is probably the simplest but adds duplication. (2) avoids duplication but requires an additional parsing of TLV type and length fields. (3) may add unwanted complexity to the macros.

Thoughts? Preferences?

Reply:
(1) also adds some more (de)structuring boilerplate since the "full" TLV stream would need to be reassembled into "partial" TLV streams to pass to the `Content` conversion functions.

Reply:
I would really need to see (2) and (3) or play with it more, either sounds good to me depending on how the macros shake out. Agreed (1) sounds too boilerplate-y. Intuitively, there may be a way to break out of `decode_tlv_stream` (or new version of it) after `peek`'ing that the next type in the reader is outside of the desired range?

Reply:
See latest push with an attempt at implementing (3). It works at least with the minimal test and isn't horrible in terms of changes required to reuse `decode_tlv_stream`. Would be nice to not need the `SeekReadable` trait. Open to better naming on that.

Reply:
FYI, pushed the latest version which had some clean-ups, including making `SeekReadable`'s visibility `pub(crate)`. Also add checks failing if all bytes are not consumed a per the spec (i.e., no bytes exist outside the expected ranges).